File: 0013-roman-to-integer/solution.java
class Solution {
    public static int romanToInt(String s) {
        Map<Character, Integer> roman = new HashMap();
        roman.put('I', 1);
        roman.put('V', 5);
        roman.put('X', 10);
        roman.put('L', 50);
        roman.put('C', 100);
        roman.put('D', 500);
        roman.put('M', 1000);

        s = s.replace("IV","IIII");
        s = s.replace("IX","VIIII");
        s = s.replace("XL","XXXX");
        s = s.replace("XC","LXXXX");
        s = s.replace("CD","CCCC");
        s = s.replace("CM","DCCCC");
        
        int result = 0;
        char[] stringArr = s.toCharArray();
        for (char c: stringArr) {
            result = roman.getOrDefault(c, 0) + result;
        }
        return result;
    }
}

File: 0013-roman-to-integer/solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        d = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M":1000}
        n = len(s)
        res = 0
        for i in range(n-1,-1,-1):
            if s[i] in d:
                if i < n-1 and d[s[i]] < d[s[i+1]] :
                    res -= d[s[i]]
                else:
                    res += d[s[i]]
        return res

